nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java

```diff
@@ -0,0 +1,438 @@
+     */
+    @Override
+    public void onConfigured(AuthorizerConfigurationContext configurationContext) throws AuthorizerCreationException {
+        // Our first init step is to select the command set based on the
+        // operating system name:
+        ShellCommandsProvider commands = getCommands();
+
+        if (commands == null) {
+            commands = getShellCommandsProvider(null);
+            setCommandsProvider(commands);
+        }
+
+        // Our second init step is to run the system check from that
+        // command set to determine if the other commands will work on
+        // this host or not.
+        try {
+            ShellRunner.runShell(commands.getSystemCheck());
+        } catch (final IOException ioexc) {
+            logger.error("initialize exception: " + ioexc + " system check command: " + commands.getSystemCheck());
+            throw new AuthorizerCreationException(SYS_CHECK_ERROR, ioexc.getCause());
+        }
+
+        // With our command set selected, and our system check passed,
+        // we can pull in the users and groups:
+        refreshUsersAndGroups();
+
+        // Our last init step is to fire off the refresh threads:
+        initialDelay = getIntegerPropertyWithDefault(configurationContext, INITIAL_REFRESH_DELAY_PROPERTY, 30);
+        fixedDelay = getIntegerPropertyWithDefault(configurationContext, REFRESH_DELAY_PROPERTY, 30);
+
+        scheduler.scheduleWithFixedDelay(this::refreshUsersAndGroups, initialDelay, fixedDelay, TimeUnit.SECONDS);
+    }
+
+    private static ShellCommandsProvider getShellCommandsProvider(String osName) {
+        if (osName == null) {
+            osName = System.getProperty("os.name");
+        }
+
+        ShellCommandsProvider commands;
+        if (osName.startsWith("Linux")) {
+            logger.debug("Selected Linux command set.");
+            commands = new NssShellCommands();
+        } else if (osName.startsWith("Mac OS X")) {
+            logger.debug("Selected OSX command set.");
+            commands = new OsxShellCommands();
+        } else {
+            throw new AuthorizerCreationException(OS_TYPE_ERROR);
+        }
+        return commands;
+    }
+
+    private int getIntegerPropertyWithDefault(AuthorizerConfigurationContext context, String name, int defaultValue) {
+        final PropertyValue property = context.getProperty(name);
+        if (property != null) {
+            return property.asInteger();
+        } else {
+            return defaultValue;
+        }
+    }
+
+    /**
+     * Called immediately before instance destruction for implementers to release resources.
+     *
+     * @throws AuthorizerDestructionException If pre-destruction fails.
+     */
+    @Override
+    public void preDestruction() throws AuthorizerDestructionException {
+        try {
+            scheduler.shutdownNow();
+        } catch (final Exception ignored) {
+        }
+    }
+
+    public ShellCommandsProvider getCommands() {
```

Comment:
Is this method necessary? Do we allow external classes to consume this internal object? And if it is necessary, its naming convention should be `getCommandsProvider()` to match `setCommandsProvider()`.

Reply:
That's a good catch.  I've renamed the method in question to `getCommandsProvider`.  I've also renamed the OS commands selector method to `getCommandsProviderFromName`.